### ISSUE TITLE
Stacked pages hierarchy rebuilt on push of relative child route

### ISSUE THREAD

**Issue by markphillips100** [tomgilder/routemaster#143]
I'm trying to make sense of some page rebuilding going on.  Not sure if it's a by design thing or a bug at this point.

I've created a fork and added a new example app [here](https://github.com/markphillips100/routemaster/tree/main/example/nested_sample).  All this app provides are 4 pages, each with a button that when clicked will navigate to one of the other pages; page1->page2->page3->page4.  The navigation is by pushing a relative path and the routes are as follows:

```
        '/one/:p1': (_) {
          print("Creating MaterialPageOne");
          return const MaterialPageOne();
        },
        '/one/:p1/two/:p2': (_) {
          print("Creating MaterialPageTwo");
          return const MaterialPageTwo();
        },
        '/one/:p1/two/:p2/three/:p3': (_) {
          print("Creating MaterialPageThree");
          return const MaterialPageThree();
        },
        '/one/:p1/two/:p2/three/:p3/four/:p4': (_) {
          print("Creating MaterialPageFour");
          return const MaterialPageFour();
        },
        '/': (_) => const Redirect('/one/1'),
```

In my fork I've added some print statements so I can see what the current page stack is when routing occurs. With this configuration I only ever see a single page in the stack (during routemaster's _navigate method [here](https://github.com/markphillips100/routemaster/blob/f7c4c5fc0a38c89fb1d15d13816b226d410ed207/lib/routemaster.dart#L496)) no matter what level of nesting there is.  Consequently, I also notice the Build method of each page in the hierarchy being called, sometimes as a result of the route builder for the page being called but sometimes without.

This is an example of the console output I see from my print statements:
On startup we see the following:
```
I/flutter (27017): Creating MaterialPageOne
I/flutter (27017): PageOne build
```
Nothing wrong there. Next, tap "page-2" button to navigate to page 2 from page 1:
```
I/flutter (27017): Routemaster page stack: [0] -> /one/1
I/flutter (27017): Creating MaterialPageTwo
I/flutter (27017): PageTwo build
I/flutter (27017): PageOne build
```
Expected the page 2 create call from the route builder but not page1's rebuild.
Then from page 2 to page 3 we see:
```
I/flutter (27017): Routemaster page stack: [0] -> /one/1/two/2
I/flutter (27017): Creating MaterialPageThree
I/flutter (27017): Creating MaterialPageOne
I/flutter (27017): PageThree build
I/flutter (27017): PageOne build
I/flutter (27017): PageTwo build
```
Expected the page 3 create call from the route builder but didn't expect to see page 1 being entirely recreated.  I suspect it's related to the current page stack only containing page2?  Add to that we also see a rebuild of page 2 - page1 build is due to the page being created again.

Is this behaviour by design or a bug? I was really hoping that if I push a relative path that the previous page hierarchy would not be recreated or rebuilt.

**Comment by bdremynoel** [comment]
Issues https://github.com/tomgilder/routemaster/issues/139 and https://github.com/tomgilder/routemaster/issues/142 are mentioning this too (with other routing but resulting to same rebuild issue).

**Comment by markphillips100** [comment]
Just to clarify @bdremynoel , is your expectation like mine that the rebuild should not occur?

**Comment by bdremynoel** [comment]
Absolutely, in our cases we're using tabs, and the stack is resest so popping the current page leads to to wrong tab state.
From this, the navigation is a bit broken.

**Comment by tomgilder** [comment]
Hi @markphillips100 thanks for the report, I'll look into this soon and get back to you.

Just to confirm: is this a performance concern, or causing issues with functionality?

**Comment by markphillips100** [comment]
Definitely functional. I use BlocProvider create in the Build, sometimes with an ..add(event) to trigger bloc to load data. So having Build fire causes workflow issues for me.

I could move the provider create out of the Build and into the route builder, wrap the page itself, but as the route builder is still called for some parents this would be a bit fruitless.

Anything I can do to help with a fix?

**Comment by tomgilder** [comment]
@markphillips100 do you mean in the build of the page (as in "Creating MaterialPageThree"), or the widget's `build()` method?

**Comment by markphillips100** [comment]
When I say "Build" I mean the `build(BuildContext context)` method of the page widget itself.  I refer to the function where "Creating MaterialPageThree" is as routemaster's route builder function.

In my actual code (not the example I added) I've now moved the initialization of the page's dependent bloc instance supplied to BlocProvider from where it was in the page widget's build() method to page state and use didChangeDependencies (along with a flag to control init once only).  I supply the state's bloc instance to the BlocProvider create property.

Long story short, I'm no longer affected by the page widget build() methods being called again for parent routes.  However, I am still affected by the routemaster route builder being called again for some parent routes (which will still run my bloc initialization logic as it creates a new page).  **So the same behaviour as seen in the example when navigating from page 2 to page 3 and MaterialPageOne is recreated instead of the existing page being reused.**

**Comment by markphillips100** [comment]
@tomgilder I've updated my example slightly to closer resemble my exact circumstances.  The page 1 route builder now uses a class for arguments and passes those to PageOne.  The class asserts that the routeData contains "query1", as it would do when routing to `/one/1?query1=val1` for example.

However, on navigating relative from page 2 to page 3 (`three/3`) as before PageOne is recreated but this time it is more obvious there is a problem as the route data no longer contains the query parameter demanded by PageOne's arguments and so the assertion fails.

**Comment by markphillips100** [comment]
@tomgilder I tried your 0.9.4-dev1 fix for those other issues and it also resolves my issue.  I can confirm previously created page routes are not recreated (the route builder function is not executed again).  Page widget build() methods still get called by Flutter but I think this is acceptable given I now initialize state outside of the build method.

Confirmed your commit changes work with the example I created, and your published package works with my other real-world code.

**Comment by tomgilder** [comment]
Awesome! Will still take a look at your code to see if there's any tests I can add to make sure this doesn't break.

**Comment by markphillips100** [comment]
That's great.  Thanks for your effort.

**Comment by tomgilder** [comment]
Publishing 0.9.4 stable shortly with this fix, thanks for your help!

**Comment by markphillips100** [comment]
@tomgilder I'm wondering if there's a regression or something environmental going on but I'm seeing this issue again.  I suspect it's environmental as I'm positive 0.9.4-dev1 was working for me but have had no success on both flutter 2.2.1 and 2.2.2 with that version or the released 0.9.4.  I've also tried on main branch with no success either.

Are you able to add the following unit test and see if it passes for you please?  It fails for me on pushing the 3rd page as page 1 routebuilder is called again (when it shouldn't) thus causing the assertion in the routebuilder trying to obtain some query string data that is rightly only passed to the last route.  Same problem as the original bug basically.

```
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:routemaster/routemaster.dart';
import 'helpers.dart';

void main() {
  testWidgets("Doesn't rebuild page hierarchy when nested page route pushed on top",
      (tester) async {
    final delegate = RoutemasterDelegate(
      routesBuilder: (BuildContext context) =>
        RouteMap(routes: {
        '/one/:p1': (routeData) {
          print('Creating MaterialPageOne');
          PageOneArguments.createFromRouteData(routeData); // Assert query parameters to ensure this route is only created once.
          return const MaterialPage<void>(child: PageOne());
        },
        '/one/:p1/two/:p2': (_) {
          print('Creating MaterialPageTwo');
          return const MaterialPage<void>(child: PageTwo());
        },
        '/one/:p1/two/:p2/three/:p3': (_) {
          print('Creating MaterialPageThree');
          return const MaterialPage<void>(child: PageThree());
        },
        '/': (routeData) => const Redirect('/one/1', queryParameters: {'query1':'val1'}),
      }));
   

    await tester.pumpWidget(MaterialApp.router(
      routeInformationParser: const RoutemasterParser(),
      routerDelegate: delegate,
    ));

    expect(find.byType(PageOne), findsOneWidget);
    expect(find.byType(PageTwo), findsNothing);
    expect(find.byType(PageThree), findsNothing);

    // Push 2nd page
    expect(
      await recordUrlChanges(() async {
        delegate.push('/one/1/two/2');
        await tester.pump();
        await tester.pump(kTransitionDuration);

        expect(find.byType(PageOne), findsNothing);
        expect(find.byType(PageTwo), findsOneWidget);
        expect(find.byType(PageThree), findsNothing);
      }),
      ['/one/1/two/2'],
    );

    // Push 3nd page
    expect(
      await recordUrlChanges(() async {
        delegate.push('/one/1/two/2/three/3');
        await tester.pump();
        await tester.pump(kTransitionDuration);

        expect(find.byType(PageOne), findsNothing);
        expect(find.byType(PageTwo), findsNothing);
        expect(find.byType(PageThree), findsOneWidget);
      }),
      ['/one/1/two/2/three/3'],
    );

  });
}

class PageOneArguments {
  final String query1;

  PageOneArguments({
    required this.query1,
  });

  factory PageOneArguments.createFromRouteData(RouteData routeData) {
    assert(routeData.queryParameters['query1'] != null);

    return PageOneArguments(
      query1: routeData.queryParameters['query1']!,
    );
  }
}

```
